modin/experimental/core/execution/native/implementations/hdk_on_native/dataframe/dataframe.py

```diff
@@ -646,26 +646,19 @@ def generate_by_name(by):
 
         agg_exprs = OrderedDict()
         if isinstance(agg, str):
-            if agg == "nlargest" or agg == "nsmallest":
-                n = kwargs["agg_kwargs"]["n"]
-                if agg == "nsmallest":
-                    n = -n
-                n = LiteralExpr(n)
-                for col in agg_cols:
-                    agg_exprs[col] = AggregateExpr(agg, [base.ref(col), n])
-            else:
-                for col in agg_cols:
-                    agg_exprs[col] = AggregateExpr(agg, base.ref(col))
+            col_to_ref = {col: base.ref(col) for col in agg_cols}
+            self._add_agg_exprs(agg, col_to_ref, kwargs, agg_exprs)
```

Comment:
Please try not to make such implicit connections, it is difficult to read the code.
```suggestion
            agg_exprs = self._add_agg_exprs(agg, col_to_ref, kwargs)
```

Reply:
This function is called in the loop below and updates the existing dictionary.

Reply:
Why not move the loop into a function if it's important?

Reply:
We have the loop only in case of dict agg. In case of dict of lists, it's even nested loop.

Reply:
> We have the loop only in case of dict agg. In case of dict of lists, it's even nested loop.

Well, I don’t see any fundamental difference.

If you think these milliseconds are important, then ok.

Reply:
I don't see much difference in code readability:
```python
self._add_agg_exprs(agg, col_to_ref, kwargs, agg_exprs)
```
vs
```python
agg_exprs.update(self._get_agg_exprs(agg, col_to_ref, kwargs))
```
The first variant requires less lines of code, less memory and computations. This is a private method. The main purpose of this method is to move common parts of logic out of groupby_agg() to avoid code duplication and it should not add a performance overhead, bigger than the method call overhead.

Reply:
The first option implicitly changes the passed argument, and this is unknown without reading the function itself. It seems to me that you are looking at the code more from the C++/C code side, this behavior is not usual for Python. In case of discrepancies in the code style, I believe we should follow the general style of the entire project, some best practices and, finally, the fundamental zen of Python (`python -m this`).

Reply:
> The first option implicitly changes the passed argument, and this is unknown without reading the function itself.

This is mentioned in the first line of the function documentation. IMHO, updating mutable data structures via a utility function call, specially designed for this purpose, is a common practice for any language. Just because, in most cases, especially in a single thread execution, this is much more efficient, than creating and merging multiple temporary structures.